### ISSUE TITLE
1837 Remaining problems

### ISSUE THREAD

**Issue by erik-vos** [Rails-18xx/Rails#346]
[1] In version 1, the S5 home hex location is still done in the initial round. It works, but is not according to the rules.
**Status:** Fixed, not yet released. During testing I discovered, that this action was assigned to the previous player (S4 owner) rather than the S5 owner. Apparently, the new player's turn was not yet started. This has been fixed (perhaps the code needs to be revisited). 

[2] During testing several problems with coal company exchanges have been encountered. The code seems to offer this option only directly after a major company has floated. In practice, offering an exchange at the start of any round sometimes occurs and sometimes not, whether mandatory or optional. It has also happened automatically. Further testing is needed to sort this all out.
**Status:** fixed, should be in release 2.4.5 (this also applies to cases [3], [12] and [13]).
P.S.: I have  no info yet on what is included in either release.

[3] If, after a coal company exchange, the related major company has too many trains, a "voluntary" train discard is offered, including a "None" option, even though it is really mandatory.
**Status:** see [2].

[4]-[8]: See next comment.

[9abc] No connection between preprinted tiles on hexes K21 (U2 home) and L22. 
**Status:** Confirmed, cause still unknown. Discussed further in issue #360.

[10] Revenue rounding.
**Status:** Fixed, see issue #372 for details.

[11] Major capitalisation wrong (always 100%).
**Status:** Fixed, should be in release 2.4.5.

[12] Missing coal exchange rounds
**Status:** See [2].

[13] Forced coal merge does not give the owner a 10% share.
**Status:** Whatever was the situation in your run, it currently works fine. Should be in release 2.4.5.

[14] More than two companies with same par price.
**Status:** Fixed, see #374.

[15] Share price dropping should be as in 1835, not 1830.
**Status:** Fixed, see issue #373.

_Edit:_ Added some remaining issues

[16ab] Crashes and other problems after buying the first 4-train.
**Status:** Testing past that point has not yet been started, as too many problems were found in earlier stages. Once these pre-4T problems are fixed to the extent that at least the finances are stable (so I don't have to recreate the SR4 test files again and again), further testing will start.

[17ab] Wrong share price movements for sold-out majors.
**Status:** Fixed, see #375.

[18] Different mine income values in v1 vs. v2?
**Status:** Fixed, see issue #376.

[19] Coal companies moving to different player?
**Status:** Never seen this.  May have been fixed by the overhaul of the Coal Exchange Round some time ago. Consider it fixed until any contrary report is posted.

**Comment by madoar** [comment]
We gave the game (1837) a try as well and discovered the following errors:

* the letters describing the coal companies on the map are too small. In addition they are positioned directly on top of the black dot which makes them very hard to read
* once the PD position was not factored in correctly when the prices of the companies were reduced during the initial auction round. The prices were reduced at the beginning of the turn of the player that bought the **previous** company paper
* when buying the second U1 paper no additional money was added to the treasury of the U1, i.e. the treasury remained at 90
* income split of the coal companies appears to be wrong. The income is split 50/50 which is not correct

**Comment by erik-vos** [comment]
I have taken the liberty to number your issues as well.

Which version did you play? (The default is the original version 1, called "Basegame").

[4] This problem was already reported by JD Galt in issue #329. The background is, that these names are not part of the tile image, but printed on top of it by code that does not know how each tile looks like. We could add a position parameter, but I prefer to wait until it is clear whether or not we can obtain a background map image.

[5] I'll investigate that. A saved file would help.

[6] I'll check that too.

[7] Why would 50/50 not be correct? Yes, I know how to create a case supporting 75/25, but do the rules really intend to say that? 
The point is, that the rules do say that the player papers are either 50% or 2x25% shares, but nowhere it is said who owns the remaining 50%. So I thought it meant to say that the share owners get 50% or 2x25% of the _total_ revenue, rather than the _split_ revenue.
If we would assume that the remaining 50% is owned by the coal company, and that the _split_ revenue would be divided between the shareholders, the company would get 75% and the players either 25% or 2x12.5%. I find it hard to believe that.
Disclaimer: I have never played 1837, I just try to close-read the rules.

[8] There we go. The 1837-specific tiles come from a different source (I think from @neutronc's Inkscape) than the older generic tiles (from TileDesigner), so small differences don't surprise me. To fix this, one of us would have to recreate these tiles (and perhaps a lot more, there are more games of which the specific tiles are currently missing in TileDesigner).

**Comment by erik-vos** [comment]
Sorry, I am wrong on [7], I was confusing coal and minor companies. Rewind.

**Comment by madoar** [comment]
Yes we played the "Basegame" variant.

[5] sadly I don't have a save game. I sometimes see a similar effect on 1835, but for 1835 it is possible that it only occurs when turns are reverted (not sure)

[7]  it was my first game of 1837 and I personally do not own the game. I heard that the income of coal companies needs to be split differently. In detail the income consists of two parts: the income of the coal mine (where the train starts) and the remaining income from the later stations. The income is then split as follows:

* the income of the coal mine is given to the company treasury
* the remaining income is split 50/50

This calculation mode seems to match the one seen in the 18xx [software by clemens](https://www.cle-mens.de/software/18xx.html).

In Rails it currently seems like both parts are added together and then split 50/50 (i.e. the owning player receives 50% of the coal mine income as well)

**Comment by neutronc** [comment]
[7] No the Coal income should be entered as direct income and that should not be split by the code in place already, but paid directly into the company treasury. At least thats how it was implemented by me earlier. Need to recheck

**Comment by neutronc** [comment]
[4] i had asked Scott for a mapimage will do so again.

**Comment by erik-vos** [comment]
[7] Just rechecked, and that's indeed how it works. In OR2, the EPP runs for 50K, of which 30K goes directly to the company, and 20K is split. So EPP gets 40K, and its owner 10K.

**Comment by jdgalt** [comment]
> [6] when buying the second U1 paper no additional money was added to the treasury of the U1, i.e. the treasury remained at 90

I believe 1837 is supposed to give every company its full capital when it floats, and that includes U1 and U3.

**Comment by madoar** [comment]
We used commit 825c06653216769aca44b3fc1c0068c53134652d to play.

[7] I have a reproduction of the issue including a save game file before and after the turn :)
In my safe game EPP has a route consisting of one coal station (30) and a small station (10).
The correct split would be 30 + 5 in the treasury of the company and 5 for the player, right?
Rails splits it 20/20 in my safe game.

It is also interesting to note that the application replaces the 30 in "Company direct income" by "0" directly after the income subturn

Safe game(s):
* [7.tar.gz](https://github.com/Rails-18xx/Rails/files/6037506/7.tar.gz)

**Comment by erik-vos** [comment]
[6] Indeed U1 and U3 (type "Minor2") have full capitalisation, and that was nicely but wrongly specified as an attribute in the `<CompanyType>` tag, where it is not even parsed. It should have been specified in its `<Capitalisation>` child tag.
Perhaps 18EU was used as a bad example: there it was confusingly specified both ways!

But even at the right place* it would not have worked, as here we have a Rails novelty. Full capitalisation as defined so far only works for companies that have a share price, and these minors have not. 

So I am now adding a new capitalisation type called "fixedCash", whereafter the initial treasury cash must be specified. Working on it.

*(In fact full capitalisation was presumed by the code, as it is the default).

[7] I'm sure the current code in the master branch does it correctly, just tried again. And that code is identical to the commit you mention. 

Your "after" saved file contains a bad SetRevenue action. Indeed it gives the result that you mention. Same thing when I do Undo then Redo. But if I do Undo then SetRevenue, the result is correct. Running the "before" saved file and then doing SetRevenue also produces the correct result.

The only reasonable conclusion I can draw, is that you, one way or another, are in fact running an older version of the code.

**Comment by madoar** [comment]
Hmm, I freshly cloned the repository before we started.
Did you enter anything manually before selecting `SetRevenue`? 

Yesterday I had the feeling that whatever is entered in `Company direct income` is set to `0` before the income is calculated.
If this is the case what would happen?
If I understand it correctly it represents the value of the Coal station right?

**Comment by neutronc** [comment]
Direct Income should be the value of the Coal mine correct

**Comment by erik-vos** [comment]
[5] So far I have not been able to reproduce that. Indeed there was a time that the PD was not correctly set, but right now it looks OK to me.

[7] I had not yet checked the spinners, so I did another test, and to my pleasant surprise both spinners were completely honoured. The dividend always turned out to be the difference between the two settable values, and the payouts were right. However, it is not yet validated that the direct revenue remains lower than the total revenue. Such a check will be added.

**Comment by erik-vos** [comment]
[6] OK, the tag `<Capitalisation type="fixedCash" cash="180"/>` now does the job for U1 and U3, at floating time, so after buying the first half share.
`

**Comment by neutronc** [comment]
> 
> 
> [5] So far I have not been able to reproduce that. Indeed there was a time that the PD was not correctly set, but right now it looks OK to me.
> 
> [7] I had not yet checked the spinners, so I did another test, and to my pleasant surprise both spinners were completely honoured. The dividend always turned out to be the difference between the two settable values, and the payouts were right. However, it is not yet validated that the direct revenue remains lower than the total revenue. Such a check will be added.

Sometimes i do code that works.. :)

**Comment by erik-vos** [comment]
A corollary to [6]: Capitalisation of the majors was also wrong: 100% in stead of the total price of the buyable shares. If the BK starts at 70K, the capital should not be 700K but 560K for the 8 buyable shares of this company. 

I had that before with the SJ in 18Scan, so the solution was available: `<Capitalisation type="part" shares="8"/>`.

**Comment by neutronc** [comment]
My bad :(

**Comment by hast18xx** [comment]
>
> 
> [7] income split of the coal companies appears to be wrong. The income is split 50/50 which is not correct
> 


I checked this issue with 2.3.1 and 2.4.0 (basic game option)
both version  do produce same result, wrong split as madoar reported.

There is a new column in the op-window  next to revenue. Its called  companyIncome and in the status window next to Rev is Direct Rev, which is editable but  without any function? or pls explain ho this should work

**Comment by erik-vos** [comment]
Just downloaded, installed and ran 2.4.0, and I see nothing wrong. In its first revenue run including 2 towns, EPP scores 50, of which 30 is direct revenue (from the mine, all for the EPP) and 20 is dividend (from the 2 towns), which is split 10/10.

The "spinners", if enabled in your configuration, are there so you can correct the total and/or the direct revenue, if you don't like any of these. The dividend is then calculated as the difference between these two values (which must be positive).

**Comment by hast18xx** [comment]
> Just downloaded, installed and ran 2.4.0, and I see nothing wrong. In its first revenue run including 2 towns, EPP scores 50, of which 30 is direct revenue (from the mine, all for the EPP) and 20 is dividend (from the 2 towns), which is split 10/10.
> 
> The "spinners", if enabled in your configuration, are there so you can correct the total and/or the direct revenue, if you don't like any of these. The dividend is then calculated as the difference between these two values (which must be positive).

yes, you are right 2.4.0 is working well now in that  case.

What I have found out so far:
a) for pre-Ungarn U2 there is no automatic route display and therefor no income see SC attached
![U2 no rev](https://user-images.githubusercontent.com/64490299/109963488-c25e0600-7cec-11eb-987e-605e6c031d02.png)

b) rounding does not work properly.  1837 rules dont give any information about that issue but in 1835 there is descrived in case of 5% prussia share that the final issue sould be rounded up in favour for the player  the next 2 SC
![Rounding U3](https://user-images.githubusercontent.com/64490299/109963801-1b2d9e80-7ced-11eb-849f-8375fad70bea.png)
![Rounding U3a](https://user-images.githubusercontent.com/64490299/109963804-1c5ecb80-7ced-11eb-91b8-149a94ee71eb.png)

c) when 50% of stock has been sold, (e.g in case of MS and BK) 100% company cash is provided instead of 80% only

d) after the end of a stock round there is no possibillity to merge a coal company in a the sspecific major comp

**Comment by jdgalt** [comment]
Re: b) - The rules for rounding payouts seem to be incomplete in some versions of the rules.  I have two -- Lonny Orgler v2.0 and an earlier Paul Work translation.  Both say that payouts to players involving half shares of U1 and U3 are rounded down.  Paul Work goes on to say that when a major company pays half, the company gets the "odd 10"; I don't think Orgler addresses that case.

**Comment by hast18xx** [comment]
> Re: b) - The rules for rounding payouts seem to be incomplete in some versions of the rules. I have two -- Lonny Orgler v2.0 and an earlier Paul Work translation. Both say that payouts to players involving half shares of U1 and U3 are rounded down. Paul Work goes on to say that when a major company pays half, the company gets the "odd 10"; I don't think Orgler addresses that case.

ok ,np with  rounding down for players income but a 50 divided in 24 for the company and 13 each for  2 players seems not okay for me, from my point of view  it should be 25, 12, 12 or 25, 13,13

**Comment by hast18xx** [comment]
more errors 1837:
Bosna Opening in green phase: it is not possible to lay a green tile. Even with map correction only yellow tiles are offered
See SC:
![Bosna opening  in green phase](https://user-images.githubusercontent.com/64490299/110005319-9fe1e200-7d18-11eb-95e1-27c3a31c557e.png)

**Comment by neutronc** [comment]
Any chance that you can upload the save game ?

**Comment by madoar** [comment]
About [7]:
I'm not sure how I could be on an "old version" when I freshly cloned the repository before building/starting rails.
I will try again over the weekend with the current master.
Are there any log messages which I could add to see what is going wrong in case I see the same issue with the (then) current master again?

**Comment by erik-vos** [comment]
a) U2 not running: I think I have seen that too in one case. To be investigated.

c) already fixed (post-2.4.0)

d), e) there were more problems with the coal exchange (see also [2] and [3]). One was, that the code was based on the original rules summary (having only the differences with 1835), that did not include the retry-before-each-round clause. The whole process is currently being reworked.

**Comment by jdgalt** [comment]
> > Re: b) - The rules for rounding payouts seem to be incomplete in some versions of the rules. I have two -- Lonny Orgler v2.0 and an earlier Paul Work translation. Both say that payouts to players involving half shares of U1 and U3 are rounded down. Paul Work goes on to say that when a major company pays half, the company gets the "odd 10"; I don't think Orgler addresses that case.
> 
> ok ,np with rounding down players income. But 50 divided in 24 for the company and 13 each for 2 players seems not okay for me, from my point of view it should be 25, 12, 12 or 25, 13,13

If U1 runs for 50 my interpretation would be 25 to company, 12 to each player (but 25 if the same player owns both shares).

**Comment by erik-vos** [comment]
> > ok ,np with rounding down players income. But 50 divided in 24 for the company and 13 each for 2 players seems not okay for me, from my point of view it should be 25, 12, 12 or 25, 13,13
> 
> If U1 runs for 50 my interpretation would be 25 to company, 12 to each player (but 25 if the same player owns both shares).

My preference would be to keep the total revenue at 50, and divide that into either 26/12/12 or 24/13/13.

**Comment by hast18xx** [comment]
> Any chance that you can upload the save game ?


[1837_20210305_0719_OperatingRound 3.1y.zip](https://github.com/Rails-18xx/Rails/files/6088771/1837_20210305_0719_OperatingRound.3.1y.zip)

**Comment by hast18xx** [comment]
there is a possibility to lay a yellow tile on Maribor (J16) in yellow, what should not be the case
![M2](https://user-images.githubusercontent.com/64490299/110090213-52588a00-7d97-11eb-8cb3-42668653e7c3.png)


![Marburg yellow](https://user-images.githubusercontent.com/64490299/110089391-5d5eea80-7d96-11eb-80a0-797cd259ea49.png)

**Comment by hast18xx** [comment]
it is not possible to upgrade tile L22 in green phase
![L22 green upgrade](https://user-images.githubusercontent.com/64490299/110093299-d9f3c800-7d9a-11eb-8822-88b67c9aae4e.png)

**Comment by hast18xx** [comment]
no  tile lay possible for Bosna

![bosna_end](https://user-images.githubusercontent.com/64490299/110094939-ab76ec80-7d9c-11eb-87d2-2730b6593a09.png)
[1837_20210305_0919_OperatingRound 3.2bosnaend.zip](https://github.com/Rails-18xx/Rails/files/6089464/1837_20210305_0919_OperatingRound.3.2bosnaend.zip)

**Comment by hast18xx** [comment]
u2 tile laying no possibillity
![u2 tile laying](https://user-images.githubusercontent.com/64490299/110099476-a2d4e500-7da1-11eb-8797-30ce01f3e47f.png)

**Comment by hast18xx** [comment]
the right value of the Kohlebahn counts from Phase 3 on which is after the first 5 train was bought and brown tiles will be possible
![Kohlebahnvalue](https://user-images.githubusercontent.com/64490299/110133003-0e33ac80-7dcc-11eb-85ef-793acf5a53ac.png)

there is no way to pay out half the dividend?
![Uploading halfpayout.png…]()

**Comment by hast18xx** [comment]
could it be that founding of Suedbahn - after first 4train,, K&K, and Ungarn is not implemented yet?

**Comment by hast18xx** [comment]
it is SR end and all shares are in players hand 
in case of 50%+ it is ok (up and one to the left)
in case of 40%  the price remains but should move up and one to the right

![SR end shift](https://user-images.githubusercontent.com/64490299/110147172-a4230380-7ddb-11eb-8d08-2b3c0dc195dd.png)

**Comment by hast18xx** [comment]
game stops, no furthrt action possible

[1837_20210305_1704_OperatingRound 7.1.zip](https://github.com/Rails-18xx/Rails/files/6092046/1837_20210305_1704_OperatingRound.7.1.zip)
![game_stop](https://user-images.githubusercontent.com/64490299/110148883-a71ef380-7ddd-11eb-8667-7cc5df5d813f.png)

**Comment by hast18xx** [comment]
now tile L22 works fine

![U2 build ok](https://user-images.githubusercontent.com/64490299/110152400-f9621380-7de1-11eb-82b4-18537f4ad724.png)
[1837_20210305_1719_OperatingRound 6.2b.zip](https://github.com/Rails-18xx/Rails/files/6092241/1837_20210305_1719_OperatingRound.6.2b.zip)

**Comment by hast18xx** [comment]
5 train was bought, brown tiles already laid but BK is not allowed to do

[1837_20210305_1751_OperatingRound 7.1a.zip](https://github.com/Rails-18xx/Rails/files/6092361/1837_20210305_1751_OperatingRound.7.1a.zip)

![brownphase](https://user-images.githubusercontent.com/64490299/110154159-362f0a00-7de4-11eb-9bb4-6b9a919a6e46.png)

**Comment by neutronc** [comment]
Many thanks for your efforts.

**Comment by hast18xx** [comment]
> > > Re: b) - The rules for rounding payouts seem to be incomplete in some versions of the rules. I have two -- Lonny Orgler v2.0 and an earlier Paul Work translation. Both say that payouts to players involving half shares of U1 and U3 are rounded down. Paul Work goes on to say that when a major company pays half, the company gets the "odd 10"; I don't think Orgler addresses that case.
> > 
> > 
> > ok ,np with rounding down players income. But 50 divided in 24 for the company and 13 each for 2 players seems not okay for me, from my point of view it should be 25, 12, 12 or 25, 13,13
> 
> If U1 runs for 50 my interpretation would be 25 to company, 12 to each player (but 25 if the same player owns both shares).

here the way as our famous ancient "lemmy" did it:
![rounding](https://user-images.githubusercontent.com/64490299/110173208-8dda6f00-7dfe-11eb-87f5-47f73ac20f92.png)
25/12/!2
(only a proposal)

**Comment by jdgalt** [comment]
> it is SR end and all shares are in players hand
> in case of 50%+ it is ok (up and one to the left)
> in case of 40% the price remains but should move up and one to the right
> 
> ![SR end shift](https://user-images.githubusercontent.com/64490299/110147172-a4230380-7ddb-11eb-8d08-2b3c0dc195dd.png)

This seems to me to be a contradiction.

Remember that the actual published game uses a hex grid for the stock market (photo attached).  The translation of that into a square grid seems to be that the upper rows are shifted a half-space to the left per row.

Thus you're correct that the movement for "all sold out, president has 50% or 60%" is up-and-left (diagonal) in both the hex grid and the square grid versions.  But the movement for "all sold out, president has 40% or less", which was up-and-right in the hex grid version, is straight up in the square grid version.

![1837_market_snapshot](https://user-images.githubusercontent.com/9399882/110176287-e9353e80-7db7-11eb-916f-5211838427fd.jpg)

**Comment by hast18xx** [comment]
> > it is SR end and all shares are in players hand
> > in case of 50%+ it is ok (up and one to the left)
> > in case of 40% the price remains but should move up and one to the right

> 
> This seems to me to be a contradiction.
> 
> Remember that the actual published game uses a hex grid for the stock market (photo attached). The translation of that into a square grid seems to be that the upper rows are shifted a half-space to the left per row.
> 
> Thus you're correct that the movement for "all sold out, president has 50% or 60%" is up-and-left (diagonal) in both the hex grid and the square grid versions. But the movement for "all sold out, president has 40% or less", which was up-and-right in the hex grid version, is straight up in the square grid version.


> > in case of 50%+ it is ok (up and one to the left)    this is ok
> > in case of 40% the price remains but should move UP
![lemmy1](https://user-images.githubusercontent.com/64490299/110201632-63bb9800-7e64-11eb-8d7a-57a88d0d9619.PNG)
![lemmy2](https://user-images.githubusercontent.com/64490299/110201634-68804c00-7e64-11eb-9942-e11a7ab6e0c7.PNG)

**Comment by hast18xx** [comment]
Kohlebahn values should be corrected (because my version1 map says so, I hope this is still valid)
E11 EPP is 30,50  should be  20,40
K31 SPB is 20,50 should be 30,50

**Comment by hast18xx** [comment]
after the first 3 train has been bought all the names on the map of the privat lines are gone

![namen weg](https://user-images.githubusercontent.com/64490299/110204538-daac5d00-7e73-11eb-9750-e6e068fee20a.png)


![namen weg1](https://user-images.githubusercontent.com/64490299/110204544-e26c0180-7e73-11eb-8f1e-4b04f6f532ae.png)

**Comment by neutronc** [comment]
> 
> 
> after the first 3 train has been bought all the names on the map of the privat lines are gone
> 
> ![namen weg](https://user-images.githubusercontent.com/64490299/110204538-daac5d00-7e73-11eb-9750-e6e068fee20a.png)
> 
> ![namen weg1](https://user-images.githubusercontent.com/64490299/110204544-e26c0180-7e73-11eb-8f1e-4b04f6f532ae.png)

Thats because the names of the privates on the map states that the hex is blocked by that private which stops to be the case after the first 3 train has been bought.

**Comment by neutronc** [comment]
> 
> 
> Kohlebahn values should be corrected (because my version1 map says so, I hope this is still valid)
> E11 EPP is 30,50 should be 20,40
> K31 SPB is 20,50 should be 30,50

I'll check with my copy and the v2 map.

**Comment by hast18xx** [comment]
> > after the first 3 train has been bought all the names on the map of the privat lines are gone

> Thats because the names of the privates on the map states that the hex is blocked by that private which stops to be the case after the first 3 train has been bought.

ok, understood, but newbies will loose the info to find the hex where the discount works

**Comment by hast18xx** [comment]
my rulebook (version1) says: chapter XII page 11


phase 2 starts with first 3Train
phase 3 starts with first 5 train

![phase2](https://user-images.githubusercontent.com/64490299/110214723-3a236080-7ea6-11eb-9a7e-460ce0007d97.png)

**Comment by erik-vos** [comment]
Yes, but the version 2 rulebook follows the current convention to name the phases after the current train type. And so do we.

**Comment by neutronc** [comment]
> 
> 
> > Kohlebahn values should be corrected (because my version1 map says so, I hope this is still valid)
> > E11 EPP is 30,50 should be 20,40
> > K31 SPB is 20,50 should be 30,50
> 
> I'll check with my copy and the v2 map.

   <IfOption name="Variant" value="Basegame">
          <Hex name="E11" tile="-37001" value="30,50" orientation="5" city="EPP" mutexId="Mine"/>
        </IfOption>
		<IfOption name="Variant" value="Romoth,1837-2ndEd.">
    	     <Hex name="E11" tile="-37001" value="20,40" orientation="5" city="EPP" mutexId="Mine"/>
        </IfOption>

**Comment by neutronc** [comment]
So its a difference in variants..

**Comment by hast18xx** [comment]
It is end of SR, MS and BK all shares were sold.
After mergingf all of the Kohlebahnen only Player "4" did receive the merged shares. He did not own a single one before


![before merge](https://user-images.githubusercontent.com/64490299/110238077-bcac2e80-7f3f-11eb-8ddb-b09d04c56d6c.png)


![merge done](https://user-images.githubusercontent.com/64490299/110238116-009f3380-7f40-11eb-984c-d14fd22a5534.png)

![merge log](https://user-images.githubusercontent.com/64490299/110238186-625f9d80-7f40-11eb-872b-46af0389d556.png)

[1837_20210307_1118_OperatingRound 4.1h.zip](https://github.com/Rails-18xx/Rails/files/6097008/1837_20210307_1118_OperatingRound.4.1h.zip)

**Comment by erik-vos** [comment]
> It is end of SR, MS and BK all shares have been sold.
> After merging is done all of the Kohlebahnen moved to Player "4". He did not own a single one before

A major update of the Coal company mergers is underway. Let's look again after that update has been released.

**Comment by neutronc** [comment]
i'll look at the map and tile problems.

**Comment by neutronc** [comment]
Ok i see the problem with the direct income being splitted what it shouldnt...

**Comment by jdgalt** [comment]
> ![merge log](https://user-images.githubusercontent.com/64490299/110238186-625f9d80-7f40-11eb-872b-46af0389d556.png)

Maybe it's a nit, but I think these messages would make more sense if the holding location for shares reserved for trade-in were called "Trade-In" rather than "Unavailable".

**Comment by neutronc** [comment]
> 
> 
> Ok i see the problem with the direct income being splitted what it shouldnt...

@erik-vos : 
` boolean hasDirectCompanyIncomeInOR
                = gameUIManager.getGameParameterAsBoolean(GuiDef.Parm.HAS_SPECIAL_COMPANY_INCOME);`

i cant find the definition anywhere ?

**Comment by erik-vos** [comment]
> Maybe it's a nit, but I think these messages would make more sense if the holding location for shares reserved for trade-in were called "Trade-In" rather than "Unavailable".

I agree that "Unavailable" is unclear, but I would prefer "Reserved". 

The Bank portfolio currently named "Unavailable" is used for two purposes:
1. to hold all certificates of a company that as a whole is not yet available (i.e. buyable in a stock round), such as the later companies in 1835.
2. to hold the certificates of an available company that are not buyable, but reserved for trade-in.

"Reserved" might be a term that best covers both purposes.

Another aspect is that the term "unavailable" in the company XML is OK for whole companies, but for certificates can perhaps better be replaced by "reserved". See the recent wiki page on Shares and Certificates where I had to make effort to distinguish the current two meanings of the "unavailable" XML attribute value. The only thing that withholds me from changing this, is that there are more urgent issues.

**Comment by erik-vos** [comment]
> 
> 
> > Ok i see the problem with the direct income being splitted what it shouldnt...
> 
> @erik-vos :
> ` boolean hasDirectCompanyIncomeInOR = gameUIManager.getGameParameterAsBoolean(GuiDef.Parm.HAS_SPECIAL_COMPANY_INCOME);`
> 
> i cant find the definition anywhere ?

It's defined in GuiDef, but I can't find right now where its value is assigned. But must be set to true somewhere, otherwise the columns Dividend and Direct Company Income in ORPanel would not appear.

**Comment by erik-vos** [comment]
> The only thing that withholds me from changing this, is that there are more urgent issues.

Oh, and the fact that many saved files may become unloadable.

**Comment by erik-vos** [comment]
[4] One thing we could do to make the coal company names on the mine hexes better visible (so far these are displayed) is to replace the towns with square cities. TileDesigner has that option (via Edit | Tile Style - but beware: it affects _all_ tiles, so the SVG export must be kept separate). If we don't change the XML, it should not have side effects.

**Comment by erik-vos** [comment]
Again I have numbered later added cases (not yet complete). In the first two comments I will try to maintain the status of each case. A few cases have already been delegated to other #issues.

**Comment by jdgalt** [comment]
One more item, and I'm not sure if it deserves a separate issue, it may not even be unique to 1837:
When using the menu to view available tiles, some "tiles" with negative numbers that are preprinted hexes are appearing in the list.  And some even appear as unlimited when using the option to make plain yellow track unlimited.

Also, it would be nice if you could find out, maybe by hovering over a tile on that screen (not the map), what are the allowed upgrades for each tile.

**Comment by jdgalt** [comment]
> there is a possibility to lay a yellow tile on Maribor (J16) in yellow, what should not be the case
> ![M2](https://user-images.githubusercontent.com/64490299/110090213-52588a00-7d97-11eb-8cb3-42668653e7c3.png)
> 
> ![Marburg yellow](https://user-images.githubusercontent.com/64490299/110089391-5d5eea80-7d96-11eb-80a0-797cd259ea49.png)
> 
> no green tiles are offered
> ![Marburg](https://user-images.githubusercontent.com/64490299/110118550-d112ee80-7dba-11eb-8038-5a8f0b20bf44.png)

I tried this again and got tiles 5 and 57 just as you did.  Apparently 2.4.5 shows 402 (yellow X tile) as the upgrade for J16 when viewing it during the initial stock round, but not when you want to upgrade the hex during a game.

When tried in green phase I am still offered tiles 5 and 57.  Apparently Rails is convinced that the preprinted J16 is white.

**Comment by erik-vos** [comment]
Already fixed. See my status note in the top comment of issue #362, case [15]. As usual, it takes some time from fixing the code to publishing a new version.

Maribor/Marburg was yellow, but the per-tile upgrade specification in `TileSet.xml` was wrong: as if it was white. There is no default rule that says that yellow must always be upgraded to green. 1856 has yellow-yellow upgrades.

In #362 [15] you reported the same case earlier in slightly different words. The case numbering and status notes at the top reflect how I try to keep track of the per-case status in never-ending issues like #362 and this one.

**Comment by jdgalt** [comment]
Sorry Erik, I did look for the earlier thread but couldn't find it from here.

**Comment by neutronc** [comment]
if you got spare time jdg i can push a 2.4.6 with eriks current fixes for you to test ?

**Comment by hast18xx** [comment]
Me too,   pls

**Comment by jdgalt** [comment]
> if you got spare time jdg i can push a 2.4.6 with eriks current fixes for you to test ?

I downloaded and installed the .deb file that was pushed, or I thought I did, but trying to play it still says it is 2.4.5+ and it still shows the S2 home hex upgrading to 5,57.   Did I mess up or is that what was pushed?

**Comment by neutronc** [comment]
I am waiting for a fix from Erik for 18EU before releasing 2.4.6. Sorry

**Comment by erik-vos** [comment]
Sorry, I had it ready, but got distracted. Now pushed directly to upstream/master, together with the ConfigWindow improvement (#386, still open).

**Comment by jdgalt** [comment]
Some results today from an hour's worth of trying 2.4.6+:  (Ubuntu 17.10, 64-bit)

[1]  Fixed.  The choice is now made at the beginning of S5's first turn.

[5]  I'm not sure whether this is still happening but I can see why it's hard to diagnose -- if a player can't afford to buy anything during his turn in a start round, his turn is silently skipped and this can happen several times in a row.  (I would like to see the same thing happen in stock rounds if the player can't buy or sell anything -- which often happens several times before all of the KK and Ug minor companies have been bought up.)

[6]  Full capitalization now works for U1 and U3, this was fixed in 2.4.5+.

[7]  Not fixed.  Coal mine income still goes into the run and is split 50-50.  There is a spinner for "company direct income" but any number I enter there has no effect whatever.  Oddly there isn't any spinner for corrections to the regular run revenue.

[10]  Fixed in 2.4.6+, for coal companies, but the fact it applies to them shows that [7] is still broken.

Two unnumbered items I noticed:

*  Italy now has the blue-green hexes.  Nice.

*  S2 home hex now shows the correct upgrades.

**Comment by erik-vos** [comment]
> [7] Not fixed. Coal mine income still goes into the run and is split 50-50. There is a spinner for "company direct income" but any number I enter there has no effect whatever. Oddly there isn't any spinner for corrections to the regular run revenue.

Well, the unbelievable fact remains that my own findings are opposite, both when running Rails from the IDE and from the 2.4.6 release (on Windows 10). Only the calculated _dividend_ is split, using the spinner _has_ the intended effect, and there _is_ a total revenue spinner that works.

So I cannot help further with this problem. What do other people experience?

**Comment by hast18xx** [comment]
I can not see any error

**Comment by hast18xx** [comment]
I can not see any error

![coalmine work](https://user-images.githubusercontent.com/64490299/113827435-98807f00-9783-11eb-8a44-b8f801404631.jpg)

**Comment by erik-vos** [comment]
Could it be that there is something wrong with the Linux installment? It looks as if it is way behind the current code base.

**Comment by neutronc** [comment]
No i think Jd is using an old save game.... :)

**Comment by jdgalt** [comment]
> No i think Jd is using an old save game.... :)

It's a new game started after upgrading.  But I did save it afterwards.  Here.
[1837_20210406_1947_OperatingRound 2.1.zip](https://github.com/Rails-18xx/Rails/files/6272534/1837_20210406_1947_OperatingRound.2.1.zip)

**Comment by jdgalt** [comment]
I did two more playtests this weekend (of 2.4.6+) and have more things to report.  I'll start numbering new items at [20] so that we don't have one number meaning two things in the same issue.

[7] I confirm Erik's finding that coal mine revenues are now correct if spinners are enabled in the configuration tab.

[14] The limit on reusing par values is now enforced.  Fixed.

[15] Share price now drops only one space when multiple shares are sold.  Fixed.

[20] At start of first stock round I notice weirdness in the net worths.  Private and minor companies all seem to be valued at face (par), but coal companies are worth zero.  This seems wrong to me; I would make them all face value (100).

[21] A new company sold out in the stock round in which it formed.  The price change is correct and reported correctly, but twice.

   All players have passed, end of SR 5
   CL price goes from 75K(C6) to 77K(B5)
   CL is sold out, price rises from 75K(C6) to 77K(B5)

[22] Once major companies are floated their stock chart tokens cover the prices so it's hard to read them.  How about letting the player drag the tokens aside to look?  Or just upper-left-justifying the prices in the squares?  (I did find a workaround: resize the stock chart window so the squares are bigger.)

[23] I built SPB to connect with its eventual parent, SB, so it did not have a legal run on OR 2.1.  I find it interesting that the payout column then showed "Withhold".  Rails is smart enough to distinguish the case where a company had no train, but not the case where it had a train but no legal route?  I propose that it should say "No route".  (Granted, I can draw that inference anyway since coal companies don't have the choice to voluntarily withhold.)

[24] When a player is being asked whether to voluntarily merge a coal company, the major company's price gets highlighted in yellow on both the Game Status window and the map.  This is helpful, except that it isn't removed when the merge round ends.

[25] When merging coal companies, if I merge one and click on Done, I don't get the chance to immediately merge another.  This can be a surprise and I don't think an undo is possible.

**Comment by jdgalt** [comment]
You're right, it's 2.4.6.  My system is still having problems rebuilding the app.  When I get a free weekend I'll back it up and reinstall.

**Comment by erik-vos** [comment]
There were various issues with wrong token positions on multiple-city tiles. These all were related to changing station numbers per stop when upgrading a tile (station numbers are defined per tile). 

This was a general problem, not limited to 1837, although Wien provided an unusual complex example. In such cases, the method `MapHex.getRelatedStop()` can provide wrong results. There already was some kludgy code trying to fix this, but that did not work for Wien. 

After fixing station mapping at such tile lays, I think this issue has now been generically fixed (no guarantees, though). At the end I could throw away that kludgy code, and restore the old, simple code. The saved files of five test cases of other games (1830, 1856 and 1880) had to be corrected as well, where these had wrong station numbers.

Another required change was to make company home city numbers (on their home hex) adaptable; this is now a state attribute of each `PublicCompany`. It will be updated every time a company home station number changes value.
